gcp_variant_transforms/bq_to_vcf.py

```diff
@@ -164,30 +173,57 @@ def _bigquery_to_vcf_shards(
+  project_id, dataset_id, table_id = bigquery_util.parse_table_reference(
+      known_args.input_table)
+  bq_variant_source = bigquery.BigQuerySource(query=query,
+                                              validate=True,
+                                              use_standard_sql=True)
   annotation_names = _extract_annotation_names(schema)
+
+  base_table_id = table_id[:table_id.find(TABLE_SUFFIX_SEPARATOR)]
+  sample_query = _SAMPLE_INFO_QUERY_TEMPLATE.format(PROJECT_ID=project_id,
+                                                    DATASET_ID=dataset_id,
+                                                    BASE_TABLE_ID=base_table_id)
+  bq_sample_source = bigquery.BigQuerySource(query=sample_query,
+                                             validate=True,
+                                             use_standard_sql=True)
   with beam.Pipeline(options=beam_pipeline_options) as p:
     variants = (p
-                | 'ReadFromBigQuery ' >> beam.io.Read(bq_source)
+                | 'ReadFromBigQuery ' >> beam.io.Read(bq_variant_source)
                 | bigquery_to_variant.BigQueryToVariant(annotation_names))
+    sample_table_rows = (
+        p
+        | 'ReadFromSampleTable' >> beam.io.Read(bq_sample_source))
     if known_args.sample_names:
-      sample_ids = (p
-                    | transforms.Create(known_args.sample_names,
-                                        reshuffle=False)
-                    | beam.combiners.ToList())
+      hash_table = (
+          sample_table_rows
+          | 'SampleNameToIdDict' >> sample_mapping_table.SampleNameToIdDict())
+      sample_names = (p
+                      | transforms.Create(known_args.sample_names,
+                                          reshuffle=False))
+      sample_ids = (sample_names
+                    | 'GetSampleIds' >>
+                    sample_mapping_table.GetSampleIds(
+                        beam.pvalue.AsSingleton(hash_table))
+                    | 'CombineSampleIds' >> beam.combiners.ToList())
+      sample_names = sample_names | beam.combiners.ToList()
     else:
+      hash_table = (
+          sample_table_rows
+          | 'SampleIdToNameDict' >> sample_mapping_table.SampleIdToNameDict())
       sample_ids = (variants
                     | 'CombineSampleIds' >>
                     combine_sample_ids.SampleIdsCombiner(
                         known_args.preserve_sample_order))
-    # TODO(tneymanov): Add logic to extract sample names from sample IDs by
-    # joining with sample id-name mapping table, once that code is implemented.
-    sample_names = sample_ids
+      sample_names = (sample_ids
+                      | 'GetSampleNames' >>
+                      sample_mapping_table.GetSampleNames(
+                          beam.pvalue.AsSingleton(hash_table))
+                      | 'CombineSampleNames' >> beam.combiners.ToList())
+      sample_ids = sample_ids | beam.combiners.ToList()
+
     _ = (sample_names
```

Comment:
Hmmm, here we need to pass a pcollection to the `ParDo` transform. However, at this point `sample_names` object is a list. I am wondering how this code didn't fail... or I am missing something here?!

Reply:
It's not a list, it's never a list until pipeline is running - it's a pcollection of a single element - list of sample names.

What happens here is over each element in the above pcollection beam runs write_vcf_header_with_sample_names method, which writes `#CHROM....` stuff and then appends the values in that particular element. Since Pcollection only has 1 element, only 1 `#CHROM...` line will be written which will have all the sample names.